src/sig_stfl/lms/sig_stfl_lms_functions.c

```diff
@@ -0,0 +1,813 @@
+	 */
+	b_ret = hss_generate_private_key(
+	            LMS_randombytes,
+	            oqs_key_data->levels,
+	            oqs_key_data->lm_type,
+	            oqs_key_data->lm_ots_type,
+	            NULL, //File handler function?
+	            oqs_key_data->sec_key,
+	            oqs_key_data->public_key, len_public_key,
+	            oqs_key_data->aux_data, oqs_key_data->len_aux_data,
+	            NULL);
+	if (b_ret) {
+		memcpy(pk, oqs_key_data->public_key, len_public_key);
+		sk->secret_key_data = oqs_key_data;
+	} else {
+		OQS_MEM_secure_free(oqs_key_data->sec_key, sk->length_secret_key * sizeof(uint8_t));
+		OQS_MEM_insecure_free(oqs_key_data->aux_data);
+		OQS_MEM_insecure_free(oqs_key_data);
+		oqs_key_data = NULL;
+		return -1;
+	}
+
+	/* TODO: store key pair, file handler */
+
+	ret = 0;
+	return ret;
+}
+#endif
+
+#ifndef OQS_ALLOW_LMS_KEY_AND_SIG_GEN
+int oqs_sig_stfl_lms_sign(UNUSED OQS_SIG_STFL_SECRET_KEY *sk, UNUSED uint8_t *sm, UNUSED size_t *smlen,
+                          UNUSED const uint8_t *m, UNUSED size_t mlen) {
+	return -1;
+}
+#else
+int oqs_sig_stfl_lms_sign(OQS_SIG_STFL_SECRET_KEY *sk,
+                          uint8_t *sm, size_t *smlen,
+                          const uint8_t *m, size_t mlen) {
+
+	size_t sig_len;
+	bool status;
+	uint8_t *sig = NULL;
+	uint8_t *priv_key = NULL;
+	oqs_lms_key_data *oqs_key_data = NULL;
+	struct hss_working_key *w = NULL;
+	struct hss_sign_inc ctx;
+	if (sk) {
+		oqs_key_data = sk->secret_key_data;
+		priv_key = oqs_key_data->sec_key;
+	} else {
+		return -1;
+	}
+	w = hss_load_private_key(NULL, priv_key,
+	                         0,
+	                         NULL,
+	                         0,
+	                         0);
+	if (!w) {
+		hss_free_working_key(w);
+		return 0;
+	}
+
+	/* Now, go through the file list, and generate the signatures for each */
+
+	/* Look up the signature length */
+
+	sig_len = hss_get_signature_len_from_working_key(w);
+	if (sig_len == 0) {
+		hss_free_working_key(w);
+		return 0;
+	}
+
+	sig = malloc(sig_len);
+	if (!sig) {
+		hss_free_working_key(w);
+		return -1;
+	}
+
+	(void)hss_sign_init(
+	    &ctx,                 /* Incremental signing context */
+	    w,                    /* Working key */
+	    NULL,                 /* Routine to update the */
+	    priv_key,       /* private key */
+	    sig, sig_len,         /* Where to place the signature */
+	    0);
+
+	(void)hss_sign_update(
+	    &ctx,           /* Incremental signing context */
+	    m,         /* Next piece of the message */
+	    mlen);             /* Length of this piece */
+
+	status = hss_sign_finalize(
+	             &ctx,               /* Incremental signing context */
+	             w,                  /* Working key */
+	             sig,                /* Signature */
+	             0);
+
+	if (!status) {
+		hss_free_working_key(w);
+		OQS_MEM_insecure_free(sig);
+		return -1;
+	}
+
+	*smlen = sig_len;
+	memcpy(sm, sig, sig_len);
```

Comment:
Is `sm` intended to be a combined signature-message object? If not, let's rename it to something else; the NIST API uses `sm` for that purpose, and it's confusing to see it refer to just a signature here.